File: Leetcode/Python3/110. Balanced Binary Tree/solution.py
# https://leetcode.com/problems/balanced-binary-tree/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = True

    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(node):
            if not node:    # If end of depth return 0
                return 0

            left, right = dfs(node.left), dfs(node.right)
            if abs(left - right) > 1:
                self.result = False
                return 0    # Already found answer so can return any number
            else:
                return max(left, right) + 1
        dfs(root)
        return self.result

File: Leetcode/Typescript/110. Balanced Binary Tree/solution.ts
// https://leetcode.com/problems/balanced-binary-tree/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
  function dfs(root: TreeNode | null): number {
    if (!root) {
      return 0;
    }

    const left = dfs(root.left);
    if (left === -1) {
      return -1;
    }

    const right = dfs(root.right);
    if (right === -1) {
      return -1;
    }

    if (Math.abs(left - right) > 1) {
      return -1;
    }

    return Math.max(left, right) + 1;
  }

  return dfs(root) !== -1;
}

function isBalanced(root: TreeNode | null): boolean {
  let balance = true;

  function dfs(root: TreeNode | null): number {
    if (!root || !balance) {
      return 0;
    }

    const left = dfs(root.left),
      right = dfs(root.right);
    if (Math.abs(left - right) > 1) {
      balance = false;
    }

    return Math.max(left, right) + 1;
  }

  dfs(root);
  return balance;
}
